### ISSUE TITLE
Compatibility with elasticsearch/elasticsearch 7.4.*

### ISSUE THREAD

**Issue by AegirLeet** [babenkoivan/scout-elasticsearch-driver#297]
**Versions**

- scout-elastic-driver: v4.0.0
- laravel/framwork: v5.8.35
- php: 7.3.11
- elasticsearch: v7.4.0

**Description**

The [7.4.0 release](https://github.com/elastic/elasticsearch-php/releases/tag/v7.4.0) for *elasticsearch/elasticsearch* introduced a breaking change:

> Using a deprecated parameter is notified triggering a E_USER_DEPRECATED error (e.g. using the `type` parameter will generate a `Specifying types in urls has been deprecated` deprecation message).

See [here](https://github.com/elastic/elasticsearch-php/blob/master/src/Elasticsearch/Endpoints/Indices/PutMapping.php#L27-L29), for example.

Because of this, requests that specify `type` will cause an error to be triggered, currently making this package incompatible with *elasticsearch/elasticsearch* 7.4.*.

A simple fix would be to restrict the *elasticsearch/elasticsearch* dependency to < 7.4.0 for now.

The proper solution is to move to the typeless API supported in Elasticserach >= 7.0, see [here](https://www.elastic.co/blog/moving-from-types-to-typeless-apis-in-elasticsearch-7-0).

**Steps to reproduce**

Use this package with *elasticsearch/elasticsearch* 7.4.* and execute an `elastic:update-mapping` command. it will trigger an error.

**Comment by iget-esoares** [comment]
Hi @AegirLeet,

Issue confirmed, it was really caused by the v7.4.0 of `elasticsearc/elasticsearch` package.

@babenkoivan I'm changing constraints for this package to `7.3.*` while we don't move away from types, okay?

**Comment by iget-esoares** [comment]
@AegirLeet and everyone involved:

While the patch is not release, It's possible to workaround this issue by requiring explicitly the version 7.3.0 of `elasticsearch/elasticsearch` on your application:

    composer require elasticsearch/elasticsearch 7.3.*

**Comment by babenkoivan** [comment]
The fixed is merged and will appear in 4.0.1, please consider update the package.

**Comment by Vmadmax** [comment]
When will 4.0.1 be available?

**Comment by iamthefreeman** [comment]
Should this issue be closed? The PR was only a temporary work around?

We still have the core problem with: 

> Using a deprecated parameter is notified triggering a E_USER_DEPRECATED error (e.g. using the type parameter will generate a Specifying types in urls has been deprecated deprecation message).

**Comment by AegirLeet** [comment]
There seems to be an open issue #96 for removing mapping types.

**Comment by ezimuel** [comment]
@iamthefreeman what is the blocking part for using `elasticsearch-php 7.4.0`? It's the trigger_error() with E_USER_DEPRECATED? 
I'm working for releasing 7.4.1 soon with some fix for unexpected BC breaks on some endpoints class name, see https://github.com/elastic/elasticsearch-php/pull/968.

Can you test this PR to see if this will solve the issue for `scout-elasticsearch-driver'?
You can use the following `composer.json` to test this PR:
```json
{
    "config": {
        "github-oauth": {
            "github.com": "<TOKEN>"
        }
    },
    "require": {
        "elasticsearch/elasticsearch": "dev-fix/967"
    },
    "repositories": [
        {
            "type": "vcs",
            "url": "https://github.com/ezimuel/elasticsearch-php"
        }
    ]
}
```
Thanks!

**Comment by iamthefreeman** [comment]
@ezimuel Yes because `E_USER_DEPRECATED` is thrown it causes fatal errors in Laravel based applications. 

Would it be possible to move this deprecation to a phpdoc deprecation and then move the `E_USER_DEPRECATED` to `8.0`? 

![Screenshot 2019-11-26 at 08 42 38](https://user-images.githubusercontent.com/916500/69613124-e9795280-1028-11ea-9323-0bade59228c7.png)

As @AegirLeet says, there is already an issue report to move away from mapping types but it looks like it's been around for a while.

**Comment by ezimuel** [comment]
@iamthefreeman ok, I'm considering to use the [silencing operator](https://www.php.net/manual/en/language.operators.errorcontrol.php) `@trigger_error` in order to suppress the error message. This is the same approach used by [Symfony](https://symfony.com/doc/current/contributing/code/conventions.html#deprecating-code) framework. The error message is not showed with `@` but it's still possible to catch it using a custom error handler (e.g. [Custom Error Handler in Laravel 5](https://sarav.co/registering-custom-exception-handler-laravel-5/)). 
I'll update you shortly, I'm doing some testing on my side.

**Comment by ezimuel** [comment]
@iamthefreeman I just sent this PR https://github.com/elastic/elasticsearch-php/pull/973 to update all the deprecated message using `@trigger_error()`. Can you test it to see if this will solve the issue? Thanks!

**Comment by iamthefreeman** [comment]
Yup, that appears to have fixed it @ezimuel,  

I've added the following to my composer to test:
```
"config": {
    "github-oauth": {
        "github.com": "<token>"
    }
},
"repositories": [
    {
        "type": "vcs",
        "url": "https://github.com/ezimuel/elasticsearch-php"
    }
],
"require": { 
    "babenkoivan/scout-elasticsearch-driver": "4.0.0",
    "elasticsearch/elasticsearch": "dev-fix/suppress-deprecation as 7.4.0",
}   
```

**Comment by ezimuel** [comment]
@iamthefreeman cool, I'll release 7.4.1 shortly! Thanks for your support and sorry again for the issue.

**Comment by ezimuel** [comment]
@AegirLeet and @iamthefreeman I just released `elasticsearch-php` [7.4.1](https://github.com/elastic/elasticsearch-php/releases/tag/v7.4.1).

**Comment by iget-master** [comment]
@ezimuel The reason we locked to v7.3.0 was really the `trigger_error`. Since this package wasn't fully migrated to don't use `type` at all, it was raising an `E_USER_DEPRECATED`.

Now I'll test reverting to v7.4.1.

**Comment by fooger** [comment]
Hello! Please, return the package to 7.4.1 or 7.5.0. I am use ElasticSearch 7.5.0 and minimum required elasticsearch-php package is 7.4.0. Before today, I am use the scout-elasticsearch-driver v.4.0.0 (where the elasticsearch-php version is NOT locked in composer.json) and last elasticsearch-php (7.4.1/7.5.0). But today I am update the Laravel to 6.9.0, and artisan not working (fixed issue #305), so I need the last scout-elasticsearch-driver, but I can not use is.
Please, disable the locking version of elasticsearch-php in composer.json. I check, your package wih last version working good.
p.s. sorry for my english
